cpp/ql/test/query-tests/Likely Bugs/Arithmetic/SignedOverflowCheck/SignedOverflowCheck.cpp

```diff
@@ -0,0 +1,104 @@
+// Signed-comparison tests
+
+/* 1. Signed-signed comparison.  The semantics are undefined. */
+bool cannotHoldAnother8(int n1) {
+    // clang 8.0.0 -O2: deleted (silently)
+    // gcc 9.2 -O2: deleted (silently)
+    // msvc 19.22 /O2: not deleted
+    return n1 + 8 < n1; // BAD
+}
+
+/* 2. Signed comparison with a narrower unsigned type.  The narrower
+      type gets promoted to the (signed) larger type, and so the
+      semantics are undefined. */
+bool cannotHoldAnotherUShort(int n1, unsigned short delta) {
+    // clang 8.0.0 -O2: deleted (silently)
+    // gcc 9.2 -O2: deleted (silently)
+    // msvc 19.22 /O2: not deleted
+    return n1 + delta < n1; // BAD
+}
+
+/* 3. Signed comparison with a non-narrower unsigned type.  The
+      signed type gets promoted to (a possibly wider) unsigned type,
+      and the resulting comparison is unsigned. */
+bool cannotHoldAnotherUInt(int n1, unsigned int delta) {
+    // clang 8.0.0 -O2: not deleted
+    // gcc 9.2 -O2: not deleted
+    // msvc 19.22 /O2: not deleted
+    return n1 + delta < n1; // GOOD
+}
+
+bool shortShort1(unsigned short n1, unsigned short delta) {
+    // clang 8.0.0 -O2: deleted
+    // gcc 9.2 -O2: deleted
+    // msvc 19.22 /O2: not deleted
+	return n1 + delta < n1; // BAD
+}
+
+bool shortShort2(unsigned short n1, unsigned short delta) {
+    // clang 8.0.0 -O2: not deleted
+    // gcc 9.2 -O2: not deleted
+    // msvc 19.22 /O2: not deleted
+	return (unsigned short)(n1 + delta) < n1; // GOOD
+}
+
+/* Distinguish `varname` from `ptr->varname` and `obj.varname` */
+struct N {
+	unsigned short n1;
+} n, *np;
+
+bool shortStruct1(unsigned short n1, unsigned short delta) {
+	return np->n1 + delta < n1; // GOOD
+}
+
+bool shortStruct1a(unsigned short n1, unsigned short delta) {
+	return n1 + delta < n.n1; // GOOD
+}
+
+bool shortStruct2(unsigned short n1, unsigned short delta) {
+	return (unsigned short)(n1 + delta) < n.n1; // GOOD
+}
+
+struct se {
+        short xPos;
+        short yPos;
+        short xSize;
+        short ySize;
+};
+
+extern se *getSo(void);
+
+bool func1(se *so) {
+	se *o = getSo();
+	if (so->xPos + so->xSize < o->xPos // GOOD
+	    || so->xPos > o->xPos + o->xSize) { // GOOD
+    	// clang 8.0.0 -O2: not deleted
+    	// gcc 9.2 -O2: not deleted
+    	// msvc 19.22 /O2: not deleted
+		return false;
+	}
+	return true;
+}
+
+bool checkOverflow3(unsigned int a, unsigned short b) {
+  return (a + b < a);  // GOOD
+}
+
+struct C {
+  unsigned int length;
+};
+
+int checkOverflow4(unsigned int ioff, C c) {
+  // not deleted by gcc or clang
+  if ((int)(ioff + c.length) < (int)ioff) return 0; // GOOD
+  return 1;
+}
+
+#define AV_INPUT_BUFFER_PADDING_SIZE 64
+
+int overflow12(int codecdata_length) {
+	if(codecdata_length + AV_INPUT_BUFFER_PADDING_SIZE <= (unsigned)codecdata_length) { // GOOD
```

Comment:
I don't see why this should be `GOOD`. If `codecdata_length` is `INT_MAX` (0x7fffffff), then the `+` will do a signed overflow, which is undefined behaviour. Assuming that doesn't make the program explode, I think it's likely that the signed addition overflows to the bit pattern 0x8000003f. When converted back to `unsigned` (implicitly, to match the RHS of `<=`), that's _not_ smaller than `(unsigned)0x7fffffff`, so `overflow12` returns 1 as if there was no overflow.

I can observe both of these problems with the following test:

```cpp
#define AV_INPUT_BUFFER_PADDING_SIZE 64

int overflow12(int codecdata_length) {
	if(codecdata_length + AV_INPUT_BUFFER_PADDING_SIZE <= (unsigned)codecdata_length) { // GOOD
	  return -1;
	}
	return 1;
}

int main(int argc, char **argv) {
  int k = 0x7fffffff;
  if (overflow12(k) == -1)
    return 1;
  return 0;
}
```

Compile with on Clang with UBSan (the undefined-behaviour sanitizer) turned on:
```
$ clang++ -fsanitize=undefined test.cc
$ ./a.out
test.cc:4:22: runtime error: signed integer overflow: 2147483647 + 64 cannot be represented in type 'int'
$ echo $?
0
```

Reply:
I agree with your analysis, though I wonder why neither `gcc` nor `clang` delete the comparison expression (which convinced me that the expression was `GOOD`).  Either they consider the semantics to be defined, or perhaps they missed an optimization opportunity.

Reply:
I've analyzed the codegen for both `gcc` and `clang`; the code generated by `gcc` is straightforward, but `clang` is a tad clever:
```
return (n + 32 <= (unsigned)n? -1: 1);

gcc 9.2:

overflow12(int):
 lea 0x20(%rdi),%eax            ; eax = rdi + 32 = n + 32  
 cmp %eax,%edi                  ; n vs. (n + 32)
 sbb %eax,%eax                  ; if n < n + 32, eax = -1, otherwise eax = 0
 and $0x2,%eax                  ; if n < n + 32, eax = 2, otherwise eax = 0
 sub $0x1,%eax                  ; if n < n + 32, eax = 1, otherwise eax = -1
 retq 

clang 9.0.0:

overflow12(int):
 xor %eax,%eax                  ; eax = 0
 cmp $0xffffffe0,%edi           ; n vs. -32
 setb %al                       ; if n < -32, eax = 1, otherwise eax = 0
 lea (%rax,%rax,1),%eax         ; if n < -32, eax = 2, otherwise eax = 0
 add $0xffffffff,%eax           ; if n < -32, eax = 1, otherwise eax = -1
 retq  
 ```

Reply:
So I guess, the philosophical question remains:  Since both mainstream compilers appear to treat the expression as semantically valid, should we be warning users against it?  (I do agree with @jbj 's analysis, though.)

Reply:
For the time being, I've incorporated @jbj's suggestions into my last commit/push.  We can still discuss whether comparisons not deleted by compilers should receive their own warnings or not (they currently will).

As I wrote in the Jira ticket, It is my opinion that the present query (`cpp/signed-overflow-check`) subsumes `cpp/bad-addition-overflow-check` and that the latter could be safely deleted.

Reply:
Thanks for explaining the assembly output. I'm impressed with GCC's bit-twiddling to produce `1` and `-1`.

I think it was missing from my analysis that the overflow check we're discussing is designed to produce `-1` if the addition overflows when the result is interpreted as unsigned. That's why it didn't detect an overflow in my test case, but the check _will_ work as intended in the particular context of https://ffmpeg.org/doxygen/3.2/rmdec_8c_source.html, where the addition result is only used as unsigned after the check.

I don't want to equate undefined behaviour with deleted code. For example, `(int)x + 2` has undefined behaviour for two possible values of `x`, but that doesn't mean that the whole expression is deleted. It just means that the compiler will generate the most efficient code for the expression while being allowed to assume that those two values of `x` are impossible. That may or may not affect code generation on a particular combination of compiler version, CPU architecture, and optimisation level. In the present test case, it seems likely that the most efficient machine code will Just Work for all inputs even if it isn't required to.

Our definition of _false positive_ is "an alert that project owners don't think should be fixed". I'd be surprised if we find a project owner who isn't concerned about undefined behaviour in their overflow checks. If that should happen, I'd like to hear from them to understand their reasoning.

Reply:
To argue that this query subsumes `cpp/bad-addition-overflow-check`, the first step is to move the files from your new test directory into [`cpp/ql/test/query-tests/Likely Bugs/Arithmetic/BadAdditionOverflowCheck`](https://github.com/Semmle/ql/tree/master/cpp/ql/test/query-tests/Likely%20Bugs/Arithmetic/BadAdditionOverflowCheck). Then the test results will show us which cases are covered by one or both of the queries.

I agree we want to avoid overlap between the new query and the existing query, but the solution isn't necessarily to merge them. If their `*.qhelp` files should contain very different explanations, then they'll have to remain separate queries.

Reply:
The `cpp/bad-addition-overflow-check` query has only one test case:
```
$ cat BadAdditionOverflowCheck.expected
| test.cpp:3:11:3:19 | ... < ... | Bad overflow check. |
```
With our `cpp/signed-overflow-check` query, we catch the same problem, plus several others:
```
$ cat SignedOverflowCheck.actual
| SignedOverflowCheck.cpp:8:12:8:22 | ... < ... | Testing for signed overflow may produce undefined results. |
| SignedOverflowCheck.cpp:18:12:18:26 | ... < ... | Testing for signed overflow may produce undefined results. |
| SignedOverflowCheck.cpp:35:9:35:23 | ... < ... | Testing for signed overflow may produce undefined results. |
| SignedOverflowCheck.cpp:99:10:99:30 | ... <= ... | Testing for signed overflow may produce undefined results. |
| test.cpp:3:11:3:19 | ... < ... | Testing for signed overflow may produce undefined results. |
```
So I would argue that `cpp/signed-overflow-check` does subsume `cpp/bad-addition-overflow-check`, though I'm always open to test cases showing otherwise.

Reply:
If both operands of the addition have small unsigned integer types, then there won't be any undefined behavior, but it will be impossible for the addition to overflow. I think that's going to merit a separate qhelp, since both the fix and the explanation of the behavior will differ

Reply:
I'm suggesting that you _actually_ merge the two test directories and push it to this PR, regardless of whether we choose to merge the queries.

I agree with @rdmarsh2, which means that `signed-overflow-check` should ideally be adjusted so it doesn't repeat any of the results from `bad-addition-overflow-check` since the qhelp won't be appropriate for those. You'll want to add some test cases that are on the border, like `(int)(unsigned short)x + 2 < (int)(unsigned short)x`.

Reply:
Instead of adjusting `signed-overflow-check` (which may introduce false positives/negatives along the way), I think it would be appropriate to simply delete `bad-addition-overflow-check`.  The `.qhelp` for `signed-overflow-check` may need adjusting.

I can certainly migrate the `signed-overflow-check` test bits over to the `bad-addition-overflow-check` directory (as several other queries live there already).  I'll also add your test case.

Reply:
I've turned @jbj 's test case into a function:
```
bool multipleCasts(char x) {
    // clang 9.0.0 -O2: deleted
    // gcc 9.2 -O2: deleted
    // msvc 19.22 /O2: deleted
    return (int)(unsigned short)x + 2 < (int)(unsigned short)x;
}
```
As you can see, all three compilers delete the relational expression in the function, replacing it with
```
xor %eax,%eax
retq 
```
So the fact that the parameter started its life as a smaller type does not matter;.